### PR TITLE
Adding Compass node

### PR DESCRIPTION
**Pull request by angel-ayala**
**Description**
Adding the [Compass node](https://www.cyberbotics.com/doc/reference/compass) to be used as a device

**Affected Packages**
List of affected packages:
  - webots_ros2_driver
  - webots_ros2_msgs

### CODE REVIEW

**Review by ygoumaz — CHANGES_REQUESTED** [review]
Thanks for this contribution! I have some comments on the code and implementation. In addition, an entry for the enhancement should be added to the [main webots_ros2 changelog](https://github.com/cyberbotics/webots_ros2/blob/master/webots_ros2/CHANGELOG.rst) and [webots_ros2_driver changelog](https://github.com/cyberbotics/webots_ros2/blob/master/webots_ros2_driver/CHANGELOG.rst).

File: webots_ros2_driver/src/plugins/static/Ros2Compass.cpp
Comment (angel-ayala) [reply webots_ros2_driver/src/plugins/static/Ros2Compass.cpp:71]: I followed the definition of the ROS language for the node here https://www.cyberbotics.com/doc/reference/compass?tab-language=ros#wb_compass_enable
In regards to the timestep value, this is readed from the URDF file definition by the [Ros2SensorPlugin.cpp file](https://github.com/cyberbotics/webots_ros2/blob/master/webots_ros2_driver/src/plugins/Ros2SensorPlugin.cpp) instantiating the `mPublishTimestepSyncedMs` var.

File: webots_ros2_driver/src/plugins/static/Ros2Compass.cpp
Comment (angel-ayala) [reply webots_ros2_driver/src/plugins/static/Ros2Compass.cpp:88]: Yes, you are right, I will update those lines including the full size of the vector using `wb_compass_get_lookup_table_size()`

File: webots_ros2_driver/src/plugins/static/Ros2Compass.cpp
Comment (ygoumaz) [reply webots_ros2_driver/src/plugins/static/Ros2Compass.cpp:71]: The ROS definition in the Webots documentation is for ROS 1 and not ROS 2. However, I think you are doing well by adding these services to your plugin. I would add a new service called `mDisableService` that allows to disable separately the device, so that it is consistent with the rest of the API.

In the future, we should also add equivalent services to other existing plugin devices of webots_ros2_driver (gps, camera, etc.)

Comment (angel-ayala) [reply webots_ros2_driver/src/plugins/static/Ros2Compass.cpp:71]: Sure, seems ok for me. I will add this discussed changes in the following commit.

File: webots_ros2_driver/src/plugins/static/Ros2Compass.cpp
Comment (ygoumaz) [inline webots_ros2_driver/src/plugins/static/Ros2Compass.cpp:93]: No need to use `request->ask`. `GetInt` message doesn't need to have an 'ask' input argument. Moreover, `wb_compass_get_sampling_period` returns 0 if the compass is not enabled.

File: webots_ros2_driver/src/plugins/static/Ros2Compass.cpp
Comment (ygoumaz) [inline webots_ros2_driver/src/plugins/static/Ros2Compass.cpp:108]:
```suggestion
    const double *lookup_table = wb_compass_get_lookup_table(mCompass);
    for (int i = 0; i < table_size; i++) {
      int offset = i * 3;
      values.at(offset) = lookup_table[offset];
      values.at(offset + 1) = lookup_table[offset + 1];
      values.at(offset + 2) = lookup_table[offset + 2];
    }
    response->value = values;
```

We can get rid of the ask input. This is not mandatory.

File: webots_ros2_driver/src/plugins/static/Ros2Compass.cpp
Comment (ygoumaz) [inline webots_ros2_driver/src/plugins/static/Ros2Compass.cpp:82]: This service could just be of type `GetBool`. No input value is required and it should return true when done.

**Review by ygoumaz — CHANGES_REQUESTED** [review]
I have some additional requests about services. The changes introduced for the `Emitter` and `Receiver` in #656 were not correct. Services messages that are used to get values (like `GetInt` , `GetFloat` and `GetBool`) should not contain any input (e.g. `bool ask`). Services with no input arguments are totally fine.

Additionally, we should add a service to get the lookup table size to be consistent with other languages of the API.

Emitter and Receiver files should be fixed in another PR, but `.srv` files can be fixed in this PR.

Changelog files should also be filled.

File: webots_ros2_driver/src/plugins/static/Ros2Compass.cpp
Comment (angel-ayala) [inline webots_ros2_driver/src/plugins/static/Ros2Compass.cpp:80]: After the last comments I should also change this lines right? using `request->value` as the time step argument, right?
```suggestion
   wb_compass_enable(mCompass, request->value);
   mAlwaysOn = true;
   mIsEnabled = true;
   response->success = true;
```

File: webots_ros2_driver/src/plugins/static/Ros2Compass.cpp
Comment (angel-ayala) [inline webots_ros2_driver/src/plugins/static/Ros2Compass.cpp:85]:
```suggestion
    response->value = true;
```

File: webots_ros2_driver/src/plugins/static/Ros2Compass.cpp
Comment (ygoumaz) [reply webots_ros2_driver/src/plugins/static/Ros2Compass.cpp]: Yes, you are right. I missed this one. The requested value should be used and not the global timestep.

File: webots_ros2_msgs/srv/GetBool.srv
Comment (ygoumaz) [inline webots_ros2_msgs/srv/GetBool.srv:2]:
```suggestion
bool value

```

File: webots_ros2_msgs/srv/GetFloatArray.srv
Comment (ygoumaz) [inline webots_ros2_msgs/srv/GetFloatArray.srv:2]:
```suggestion
float64[] value
```

As a convention, we keep it singular.

File: webots_ros2_driver/src/plugins/static/Ros2Compass.cpp
Comment (ygoumaz) [inline webots_ros2_driver/src/plugins/static/Ros2Compass.cpp:102]:
```suggestion
    response->value = values;
```

**Review by ygoumaz — CHANGES_REQUESTED** [review]
A few minor suggestions left. Also as mentioned in my last review, an additional service for the lookup table size should be added.

File: webots_ros2_driver/src/plugins/static/Ros2Compass.cpp
Comment (lukicdarkoo) [inline webots_ros2_driver/src/plugins/static/Ros2Compass.cpp:71]: Sorry for randomly jumping into the PR, but a few notes here.

According to the documentation, these values should be in Tesla units:

> x, y, and z components of the field vector in Tesla
- http://docs.ros.org/en/noetic/api/sensor_msgs/html/msg/MagneticField.html

Therefore, we should take the raw value from Webots, calculate the actual value using the lookup table and convert those values to Teslas. As the result, the `get_lookup_table` service is irrelevant.

Other services `enable`, `disable`, and `get_sampling_period` are almost never used in real hardware. Therefore, exposing them could potentially make the sim2real transfer harder (we want to minimize the reality gap).

It would be probably useful to publish other types of messages, e.g.:
https://github.com/clearpathrobotics/imu_compass/blob/master/src/imu_compass.cpp
Maybe `sensor_msgs/Imu` with only the `orientation` field populated. Ideally, we should take a look into a few existing compass interfaces and expose only the most generic part of their ROS API.

Comment (ygoumaz) [reply webots_ros2_driver/src/plugins/static/Ros2Compass.cpp:71]: Thanks for your comments! They are really useful. I now understand why all these services are not included in other plugins. I agree that these services should therefore not be included here either. This also applies to the new Emitter and Receiver nodes. I also agree that we should be consistent with other compass interfaces, like the one you have linked.

Comment (ygoumaz) [reply webots_ros2_driver/src/plugins/static/Ros2Compass.cpp:71]: Actually, the Webots Compass node (and magnetometers in general) return a 3D vector of the North direction. In the interface linked by @lukicdarkoo, the code gets the compass measuremement in a message of type [Vector3Stamped](http://docs.ros.org/en/noetic/api/geometry_msgs/html/msg/Vector3Stamped.html). We should use this instead of the MagneticField type.

Comment (lukicdarkoo) [reply webots_ros2_driver/src/plugins/static/Ros2Compass.cpp:71]: @ygoumaz I would probably find Vector3Stamped more useful than MagneticField. Also, a heading (as Float32) would be quite useful for most use cases.

**Comment by ygoumaz** [comment]
Yes, sure! Here are the main points to modify in the new Compass node, and in the existing Emitter and Receiver ones:
- All services should be removed. As mentioned by @lukicdarkoo, they don't reflect the reality, so we don't want to add them. Also, this is why they are not implemented in other sensors plugins (GPS, etc.).
- Two separate Compass data should be published to topics:
  - The first one is the vector returned by `wb_compass_get_values() `and should be published to the topic you have already created. But instead of `MagneticField` messages type, it should be `Vector3Stamped` type.
  - The second one is not yet implemented but should by a Floating point value in degrees (or radians) representing the heading of the robot. The computation can be achieved using a simple algorithm (you should implement it in Python):
    ```c
    double get_bearing_in_degrees() {
      const double *north = wb_compass_get_values(tag);
      double rad = atan2(north[1], north[0]);
      double bearing = (rad - 1.5708) / M_PI * 180.0;
      if (bearing < 0.0)
        bearing = bearing + 360.0;
      return bearing;
    }
    ```
- All unnecessary services messages (.srv) should be removed.

The respective changes about the Compass should be performed in this PR, when changes on the Emitter and Receiver should be performed in #680.

**Comment by ygoumaz** [comment]
>   * The second one is not yet implemented but should be a Floating point value in degrees (or radians) representing the heading of the robot. The computation can be achieved using a simple algorithm (you should implement it in Python)

I am sorry, there is obviously an error in my last comment. The algorithm should not be implemented in Python. It can simply be added to the cpp plugin file of the Compass.

Once the PR is ready for a new review, you can re-request it by using the double arrow symbol on the top right of the page.

File: webots_ros2_driver/src/plugins/static/Ros2Compass.cpp
Comment (ygoumaz) [inline webots_ros2_driver/src/plugins/static/Ros2Compass.cpp:68]:
```suggestion
    const double rad = atan2(north_vector[1], north_vector[0]);
```

File: webots_ros2_msgs/srv/GetFloatArray.srv
Comment (ygoumaz) [inline webots_ros2_msgs/srv/GetFloatArray.srv:3]:
```suggestion
```

File: webots_ros2_msgs/CMakeLists.txt
Comment (ygoumaz) [inline webots_ros2_msgs/CMakeLists.txt:34]:
```suggestion
```

**Review by ygoumaz — CHANGES_REQUESTED** [review]
Some minor remaining comments.

**Review by ygoumaz — CHANGES_REQUESTED** [review]
[GetFloatArray.srv](https://github.com/cyberbotics/webots_ros2/blob/446b7c15f1e3c8e0b9768f52bba3653a98b72479/webots_ros2_msgs/srv/GetFloatArray.srv) should be completely deleted.

**Comment by ygoumaz** [comment]
I have tested the compass on the mavic example and it works nicely. This should be the last change to do before merging this PR!

**Review by ygoumaz — APPROVED** [review]
Thank you. The PR is ready to be merged. I just posted an additional minor suggestion. Feel free to apply it before merging. I believe using `north_vector` as topic name is more explicit than `values`.

**Review by ygoumaz — APPROVED** [review]

**Comment by ygoumaz** [comment]
I agree for some last changes. But a lot of changes appeared after other previous changes. Also I usually make a first code review and then test the compilation and behavior. Certain changes may get my attention only after these steps.

**Review by ygoumaz — APPROVED** [review]

**Pull request merged**
